File: crates/processing_pyo3/examples/shapes.py
from mewnala import *
from math import pi, sin, cos

t = 0.0

def setup():
    size(600, 600)

def draw():
    global t
    background(235, 230, 222)

    fill(51, 76, 127, 64)
    stroke_weight(1.5)
    stroke(38, 51, 89)

    n = 8
    begin_shape(POLYGON)
    for i in range(n + 3):
        idx = (i + n - 1) % n
        angle = idx * 2 * pi / n
        r = 100 + 30 * sin(t * 2 + idx * 0.8)
        curve_vertex(170 + r * cos(angle), 200 + r * sin(angle))
    end_shape(CLOSE)

    fill(38, 127, 76, 51)
    stroke_weight(1)
    stroke(25, 89, 51)

    pcx = 430
    pcy = 200
    s = 70 + 10 * sin(t * 1.5)

    begin_shape(POLYGON)
    vertex(pcx, pcy - s)
    bezier_vertex(pcx + s, pcy - s, pcx + s, pcy + s, pcx, pcy + s)
    bezier_vertex(pcx - s, pcy + s, pcx - s, pcy - s, pcx, pcy - s)
    end_shape(CLOSE)

    fill(204, 89, 25, 38)
    no_stroke()

    begin_shape(TRIANGLE_FAN)
    vertex(170, 450)
    for i in range(17):
        angle = i * 2 * pi / 16
        r = 60 + 20 * sin(t * 3 + angle * 2)
        vertex(170 + r * cos(angle), 450 + r * sin(angle))
    end_shape(CLOSE)

    fill(127, 51, 153, 51)
    stroke_weight(0.5)
    stroke(76, 25, 102, 102)

    begin_shape(TRIANGLE_STRIP)
    for i in range(16):
        x = 320 + i * 17
        wave = 30 * sin(t * 2 + i * 0.4)
        vertex(x, 420 + wave)
        vertex(x, 480 + wave)
    end_shape()

    t += 0.02

run()

File: examples/shapes.rs
use processing_glfw::GlfwContext;
use std::f32::consts::PI;

use processing::prelude::*;
use processing_render::render::command::DrawCommand;

fn main() {
    sketch().unwrap();
    exit(0).unwrap();
}

fn sketch() -> error::Result<()> {
    let mut glfw_ctx = GlfwContext::new(600, 600)?;
    init(Config::default())?;

    let surface = glfw_ctx.create_surface(600, 600)?;
    let graphics = graphics_create(surface, 600, 600, TextureFormat::Rgba16Float)?;

    let mut t: f32 = 0.0;

    while glfw_ctx.poll_events() {
        graphics_begin_draw(graphics)?;

        graphics_record_command(
            graphics,
            DrawCommand::BackgroundColor(bevy::color::Color::srgb(0.92, 0.90, 0.87)),
        )?;

        graphics_record_command(
            graphics,
            DrawCommand::Fill(bevy::color::Color::srgba(0.2, 0.3, 0.5, 0.25)),
        )?;
        graphics_record_command(graphics, DrawCommand::StrokeWeight(1.5))?;
        graphics_record_command(
            graphics,
            DrawCommand::StrokeColor(bevy::color::Color::srgb(0.15, 0.2, 0.35)),
        )?;

        let n = 8;
        graphics_record_command(
            graphics,
            DrawCommand::BeginShape {
                kind: ShapeKind::Polygon,
            },
        )?;

        for i in 0..n + 3 {
            let idx = (i + n - 1) % n;
            let angle = idx as f32 * 2.0 * PI / n as f32;
            let r = 100.0 + 30.0 * (t * 2.0 + idx as f32 * 0.8).sin();
            graphics_record_command(
                graphics,
                DrawCommand::ShapeCurveVertex {
                    x: 170.0 + r * angle.cos(),
                    y: 200.0 + r * angle.sin(),
                },
            )?;
        }

        graphics_record_command(graphics, DrawCommand::EndShape { close: true })?;

        graphics_record_command(
            graphics,
            DrawCommand::Fill(bevy::color::Color::srgba(0.15, 0.5, 0.3, 0.2)),
        )?;
        graphics_record_command(graphics, DrawCommand::StrokeWeight(1.0))?;
        graphics_record_command(
            graphics,
            DrawCommand::StrokeColor(bevy::color::Color::srgb(0.1, 0.35, 0.2)),
        )?;

        let pcx = 430.0;
        let pcy = 200.0;
        let s = 70.0 + 10.0 * (t * 1.5).sin();

        graphics_record_command(
            graphics,
            DrawCommand::BeginShape {
                kind: ShapeKind::Polygon,
            },
        )?;
        graphics_record_command(graphics, DrawCommand::ShapeVertex { x: pcx, y: pcy - s })?;
        graphics_record_command(
            graphics,
            DrawCommand::ShapeBezierVertex {
                cx1: pcx + s,
                cy1: pcy - s,
                cx2: pcx + s,
                cy2: pcy + s,
                x: pcx,
                y: pcy + s,
            },
        )?;
        graphics_record_command(
            graphics,
            DrawCommand::ShapeBezierVertex {
                cx1: pcx - s,
                cy1: pcy + s,
                cx2: pcx - s,
                cy2: pcy - s,
                x: pcx,
                y: pcy - s,
            },
        )?;
        graphics_record_command(graphics, DrawCommand::EndShape { close: true })?;

        graphics_record_command(
            graphics,
            DrawCommand::Fill(bevy::color::Color::srgba(0.8, 0.35, 0.1, 0.15)),
        )?;
        graphics_record_command(graphics, DrawCommand::NoStroke)?;

        graphics_record_command(
            graphics,
            DrawCommand::BeginShape {
                kind: ShapeKind::TriangleFan,
            },
        )?;
        graphics_record_command(graphics, DrawCommand::ShapeVertex { x: 170.0, y: 450.0 })?;
        for i in 0..=16 {
            let angle = i as f32 * 2.0 * PI / 16.0;
            let r = 60.0 + 20.0 * (t * 3.0 + angle * 2.0).sin();
            graphics_record_command(
                graphics,
                DrawCommand::ShapeVertex {
                    x: 170.0 + r * angle.cos(),
                    y: 450.0 + r * angle.sin(),
                },
            )?;
        }
        graphics_record_command(graphics, DrawCommand::EndShape { close: true })?;

        graphics_record_command(
            graphics,
            DrawCommand::Fill(bevy::color::Color::srgba(0.5, 0.2, 0.6, 0.2)),
        )?;
        graphics_record_command(graphics, DrawCommand::StrokeWeight(0.5))?;
        graphics_record_command(
            graphics,
            DrawCommand::StrokeColor(bevy::color::Color::srgba(0.3, 0.1, 0.4, 0.4)),
        )?;

        graphics_record_command(
            graphics,
            DrawCommand::BeginShape {
                kind: ShapeKind::TriangleStrip,
            },
        )?;
        for i in 0..16 {
            let x = 320.0 + i as f32 * 17.0;
            let wave = 30.0 * (t * 2.0 + i as f32 * 0.4).sin();
            graphics_record_command(graphics, DrawCommand::ShapeVertex { x, y: 420.0 + wave })?;
            graphics_record_command(graphics, DrawCommand::ShapeVertex { x, y: 480.0 + wave })?;
        }
        graphics_record_command(graphics, DrawCommand::EndShape { close: false })?;

        graphics_end_draw(graphics)?;
        t += 0.02;
    }

    Ok(())
}
